3_harmonize/src/harmonize_sdd.R

```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+           hour(harmonized_local_time) >= 10 & hour(harmonized_local_time) < 14) ~ 0,
+        # tag 1: time is not within 10a-2p and is not 11:59:59
+        (format(harmonized_local_time, "%H:%M:%S") != "11:59:59" &
+           (hour(harmonized_local_time) < 10 | hour(harmonized_local_time) >= 14)) ~ 1,
+        # tag 2: time is 11:59:59 or NA
+        (format(harmonized_local_time, "%H:%M:%S") == "11:59:59" | is.na(harmonized_local_time)) ~ 2,
+        .default = NA_integer_
+      ),
+      harmonized_local_time = format(harmonized_local_time, "%Y-%m-%d %H:%M:%S") # return harmonized_local_time to characters
     )
-  )
-  
-  dropped_fractions <- tibble(
+
+  # scope_tag
+  scope_text <- "scope"
+
+  negate_scope_text <- paste(
+    "no.*viewscope", "no.*scope", "without.*scope","not using scope",
+    "no scope.*used","viewscope.*not.*sent","viewscope.*not.*collected",
+    "leaky.*scope","scope.*needs.*resealing","unable.*get.*readings.*scope",
+    "hard to use.*scope","issues with viewscope","no readings.*viewscope",
+    "no.*s\\.?d\\.?.*viewscope","secchi.*not using","Viewscope: No",
+    "Viewscope used\\? - No","-Viewscope used\\? - No","Without Scope",
+    "without viewscope","no viewscope used","Viewscope: No;",
+    "scope of cable", "scope of collection", "scope of sample collection",
+    sep = "|")
+
+  comment_cols <- c("ResultAnalyticalMethod.MethodName", "ActivityCommentText",
+                    "ResultCommentText")
+
+  sdd_scope_tag <- sdd_time_tag %>%
+    # tag: 0 = Either the method/comments/p-code indicate that a viewscope was used
+    #      1 = P-code does not match viewscope, No indication that a viewscope was used, or there is negate language regarding viewscope used
+    mutate(scope_tag = case_when(
+      rowSums(across(all_of(comment_cols), ~grepl(negate_scope_text, .x, ignore.case = TRUE))) > 0  ~ 1, # negate language detected
+      rowSums(across(all_of(comment_cols), ~grepl(scope_text, .x, ignore.case = TRUE))) > 0 | USGSPCode == 72187 ~ 0, # scope language detected or correct USGS code
+      .default = 1) # if negate language is not detected, scope language is not detected and the p-code does not match the one for view scope default to 1
+    )
+
+  # tier for SDD
+  tiered_methods_sdd <- sdd_scope_tag %>%
+    mutate(tier = case_when(
+      # tier: 0 = Restrictive. Time reported within the allotted window and method indicates use of viewscope.
+      #       1 = Narrowed. Either time reported within the allotted window or method indicates use of viewscope, but not both.
+      #       2 = Inclusive. Time not reported/time reported outside of window, and no indication of viewscope used.
+      #       3 = Inclusive. harmonized value gap filled with depth column.
+      (time_tag == 0) & (scope_tag == 0) ~ 0,
+      xor((time_tag == 0), (scope_tag == 0)) ~ 1,
+      (time_tag != 0) & (scope_tag != 0) ~ 2,
+      .default = 2
+      ),
+      tier = if_else(approx_flag != 0, 3, tier)
```

Comment:
Just need to adjust from new approx_flag 1 where we use abs for negative values
```suggestion
      tier = if_else(approx_flag > 1, 3, tier)
```

Reply:
approx_flag 1 is still when the value is approximated from `ActivityDepthHeightMeasure.MeasureValue` column. The negative value conversion now falls under the misc flag. Do you want me to update this so that it a record is tier 3 if it is flagged with the misc flag indicating a special character removal?

Reply:
Superb. You can leave as you wrote it then. 

``` 
          tier = if_else(approx_flag != 0, 3, tier)
```